gdal/swig/python/osgeo/auxiliary/color_palette.py

```diff
@@ -0,0 +1,348 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# ******************************************************************************
+#
+#  Project:  GDAL
+#  Purpose:  module for handling color palettes
+#  Author:   Idan Miara <idan@miara.com>
+#
+# ******************************************************************************
+#  Copyright (c) 2020, Idan Miara <idan@miara.com>
+#
+#  Permission is hereby granted, free of charge, to any person obtaining a
+#  copy of this software and associated documentation files (the "Software"),
+#  to deal in the Software without restriction, including without limitation
+#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
+#  and/or sell copies of the Software, and to permit persons to whom the
+#  Software is furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice shall be included
+#  in all copies or substantial portions of the Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#  DEALINGS IN THE SOFTWARE.
+# ******************************************************************************
+
+import os
+import re
+from xml.dom import minidom
+from collections import OrderedDict
+import glob
+import tempfile
+from osgeo.auxiliary import base
+
+# from typing import Sequence, Union
+# from pathlib import Path
+# PathOrStrings = Union[Path, str, Sequence[str]]
+# ColorPaletteOrPathOrStrings = Union['ColorPalette', PathOrStrings]
+
+
+class ColorPalette:
```

Comment:
I'm still missing how a ColorPalette is supposed to be different from a ColorTable ? There would need to be some documentation.
I'm also a bit uneasy to expend too much the GDAL Python API from the C++ core. In particular we have no real proper Python doc and have resorted to pointing to the C++ one, but from pure Python API, we have no such possibility

Reply:
> I'm still missing how a ColorPalette is supposed to be different from a ColorTable ? There would need to be some documentation.

I needed a more generalized data structure than ColorTable.
ColorTable maps indices to a triplets/quadlets of bytes and is a thin wrapper around the c++ class. If I recall correctly all the table keys needs to be filled up even if unused.
ColorPalette is more generalized, it maps numbers/percentages/others into a 32-bit color.
I convert a ColorPalette into a ColorTable just when I need to set it into a DS.

I was about to document it if this concept is accepted.

> I'm also a bit uneasy to expend too much the GDAL Python API from the C++ core. In particular we have no real proper Python doc and have resorted to pointing to the C++ one, but from pure Python API, we have no such possibility

I see your point. 
I needed to this auxiliary class and a few more to support more functionality into gdal_calc and gdal_viewshed and others.
I think we need to find an approach to allow Python auxiliary modules to support gdal utils.
Once we migrate to Python 3 - documentation would become much easier with type hinting.
Than we could autogenerate documentation for the osgeo.* sub packages.

Reply:
@rouault What do you think about this PR?
Does the separation between osgeo.auxiliary and osgeo.gdal_auxiliary makes sense to you? 
What do you think about the ColorPalette class ? I will document if the general idea is accepted.

Reply:
> Does the separation between osgeo.auxiliary and osgeo.gdal_auxiliary makes sense to you?

I find this perhaps a bit complicated. That would make 3 subpackages with somewhat vague names: utils, auxiliary, gdal_auxiliary.

Reply:
I somewhat agree, but I couldn't come up with better names...
We could do the separation at the module level instead of at the subpackage level.

Reply:
Another idea could be to put the modules that do not depend on osgeo on their own package, outside the osgeo package.
The naming issue still stands - maybe osgeo_aux ? this package could be reused outside gdal without being dependent on gdal.
One could argue to put it outside this repo, but then we couldn't use it as auxiliary for osgeo package as dependence on outside modules is not allowed.

Reply:
Seems very complex to me... is there any _known_ use of these classes outside of GDAL? Just because it doesn't technically depend on it doesn't mean that inside GDAL's namespace isn't the right place for it to live. 

Is there a reason it can't all go into `osgeo.utils`, and if there are big worries about compatibility, put it inside `osgeo.utils.internal` with a warning saying "no compatibility guarantees — if you want it as a public API open a PR" 😄

Reply:
> Seems very complex to me... is there any _known_ use of these classes outside of GDAL? Just because it doesn't technically depend on it doesn't mean that inside GDAL's namespace isn't the right place for it to live.
> 
> Is there a reason it can't all go into `osgeo.utils`, and if there are big worries about compatibility, put it inside `osgeo.utils.internal` with a warning saying "no compatibility guarantees — if you want it as a public API open a PR" 😄

so we have `osgeo.utils` for all the `scripts` that were transformed into independent modules.
and `osgeo.auxiliary` for the common helper functions that are not functional by themselves.
I could merge `osgeo_aux` into `osgeo.auxiliary` if it lessen the clutter of naming.

Other than that what do you think about this PR?